Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Радует, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* Я оставил несколько советов, надеюсь, что они будут учтены в будущих проектах.
* Работа получилась отличной, тебе удалось добиться достаточно хорошего качества. Поздравляю!
* Проект зачтен, удачи в дальнейшем обучении!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li></ul></li><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Классификатор-LogisticRegression" data-toc-modified-id="Классификатор-LogisticRegression-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Классификатор LogisticRegression</a></span></li><li><span><a href="#Классификатор-DecisionTreeClassifier" data-toc-modified-id="Классификатор-DecisionTreeClassifier-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Классификатор DecisionTreeClassifier</a></span></li><li><span><a href="#Классификатор-SGDClassifier" data-toc-modified-id="Классификатор-SGDClassifier-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Классификатор SGDClassifier</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span><ul class="toc-item"><li><span><a href="#Итого" data-toc-modified-id="Итого-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Итого</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп» c BERT

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

<div class="alert alert-block alert-info">В связи с тем, что embeddings для 80000 записей вместо 150тыс как в оригинале даже на платном Colab на GPU считаются 1 час. То расссчитанные Embeddings сохранены в файл и в дальнейшем при реализации проекта просто загружаются уже готовые.<br>
    Опять же сначала проект делался на 2тыс записей (расчет на Colab Free 15 мин - пока манипулировал закончилось время). В таком случае даже при всех вариантах испробованной балансировки не удавалось получить f1>0.75<br>
    Позже я просчитал 8 тыс записей и модели показали нужны скоринг.<br>
    Структуру работы оставил при этом как должно было быть от начала загрузки сырых данных, их обрабротки, балансировки и работы с моделями, если бы позволяли ресурсы хоста на котором ведется работа
</div>

## Подготовка

In [3]:
#
import pandas as pd
#
pd.set_option('display.float_format', lambda x: '%.3f' % x)
pd.set_option("precision", 3)
#
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go
from plotly.offline import init_notebook_mode, iplot

sns.set_style('darkgrid')
#
import numpy as np
from datetime import datetime
from scipy import stats as st
#
from statsmodels.tsa.seasonal import seasonal_decompose
#
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, KFold, StratifiedKFold
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler, PowerTransformer
pd.options.mode.chained_assignment = None
#
from lightgbm import LGBMRegressor
#
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.linear_model import LogisticRegression, LinearRegression, Lasso, Ridge, SGDClassifier
from sklearn.dummy import DummyClassifier, DummyRegressor
#
from sklearn.metrics import accuracy_score, confusion_matrix, recall_score, precision_score, \
                            f1_score, roc_auc_score, roc_curve, mean_squared_error, mean_absolute_error, \
                            make_scorer
#
from sklearn.utils import shuffle
#
from catboost import CatBoostClassifier
#
import torch
import transformers
from tqdm import notebook
#
from IPython.display import display, HTML, display_html
#
import copy
#
import warnings
warnings.filterwarnings('ignore')

<div class="alert alert-block alert-info">
<b>Совет: </b> Желательно чтобы все импорты были собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [20]:
its_colub = True

In [21]:
%%time

#откроем датасет
if its_colub:
    df = pd.read_csv('./datasets/toxic_comments_balanced_8000.csv')
else:
    df = pd.read_csv('./datasets/toxic_comments.csv', nrows=2000)

df.name = 'toxic_comments'

print('Таблица',df.name)
#столбцы к нижнему регистру
df.columns = df.columns.str.lower()
#
display(df.head(5))
display('Размерность данных:', df.shape)
display('Типы:', df.dtypes)

Таблица toxic_comments


unnamed: 0                                               text  toxic
0       56058                     REDIRECT User talk:Pusong Bato      0
1       33422  Please feel free to add anything you feel I've...      0
2      136645  ... AmeriKKKans are fucking gay ... U.S.A. suc...      1
3      112347  Splash15hotel, ur a fucking faggot, dude. Get ...      1
4      131278  M1ss1ontomars2k4, please feel free to lick my ...      1

'Размерность данных:'

(8000, 3)

'Типы:'

unnamed: 0     int64
text          object
toxic          int64
dtype: object

CPU times: user 50.2 ms, sys: 93.3 ms, total: 144 ms
Wall time: 318 ms


<div class="alert alert-block alert-success">
<b>Успех:</b> Загрузка и первичный осмотр данных проведены хорошо.
</div>

In [5]:
def na_persentage(dat):
    print('Доля нулевых значений:')
    gaps_in_the_data = (pd
                        .DataFrame({'percent_of_nulls':dat.isna().sum()/len(dat)})
                        .sort_values(by='percent_of_nulls', ascending=False)
                       )
    gaps_in_the_data['percent_of_nulls'] = gaps_in_the_data['percent_of_nulls'].map(lambda x: '{:.2%}'.format(x))
    display(gaps_in_the_data)

na_persentage(df)
#Посчитаем количество полных дубликатов
print('Количество полных дубликатов:', df.duplicated().sum())
print()

Доля нулевых значений:


percent_of_nulls
text             0.00%
toxic            0.00%

Количество полных дубликатов: 0



In [6]:
display(df['toxic'].value_counts())
class_ratio = df['toxic'].value_counts()[0] / df['toxic'].value_counts()[1]
print(class_ratio)

0    1790
1     210
Name: toxic, dtype: int64

8.523809523809524


Классы несбалансированы. Отношение 1:8.52

In [7]:
# df_0 = df[df['toxic'] == 0]
# df_1 = df[df['toxic'] == 1]

# toxic_0_downsample = df_0.sample(4000, random_state=12345)
# toxic_1_downsample = df_1.sample(4000, random_state=12345)
# df = pd.concat([toxic_0_downsample, toxic_1_downsample])
# df = shuffle(df, random_state=12345)

# print(df.shape)

<div class="alert alert-block alert-info">
<b>Совет: </b> Напомню, что баланс классов можно изменять только в треине. В тесте и в валидации его изменять категорически запрещено, так как результаты получатся завышенными. Насколько я понял, эта ячейке не была запущена.
</div>

In [9]:
df['text'] = df['text'].values.astype('U')

## Обучение

In [10]:
%%time

has_tokenized = True

if not has_tokenized:
  tokenizer = transformers.BertTokenizer(
      vocab_file='/content/drive/MyDrive/Colab Notebooks/datasets/bert_multi_cased_L-12_H-768_A-12/vocab.txt')

  tokenized = df['text'].apply(
    lambda x: tokenizer.encode(x, add_special_tokens=True, max_length=512, truncation=True))

  max_len = max(map(len,tokenized))

  padded_tokenized = np.array([i + [0]*(max_len - len(i)) for i in tokenized])
  display(tokenized)

  attention_mask = np.where(padded_tokenized != 0, 1, 0) 

  config = transformers.BertConfig.from_json_file(
      '/content/drive/MyDrive/Colab Notebooks/datasets/bert_multi_cased_L-12_H-768_A-12/bert_config.json')
  model = transformers.BertForPreTraining.from_pretrained(
      '/content/drive/MyDrive/Colab Notebooks/datasets/bert_multi_cased_L-12_H-768_A-12/bert_model.ckpt.index', config=config, from_tf=True) 

  embeddings = []

  if torch.cuda.is_available():
      device='cuda'
      batch_size = 8
      #
      import gc
      gc.collect()
      torch.cuda.empty_cache()
      #
      model = model.to(device)
  else:
      batch_size = 25
      device='cpu'

  for i in notebook.tqdm(range(padded_tokenized.shape[0] // batch_size)):
      batch = torch.LongTensor(padded_tokenized[batch_size*i:batch_size*(i+1)]).to(torch.device(device))
      attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).to(torch.device(device))

      with torch.no_grad():
          batch_embeddings = model(batch, attention_mask=attention_mask_batch)

      if torch.cuda.is_available():
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy())
      else:
        embeddings.append(batch_embeddings[0][:,0,:].numpy())
      

  features = np.concatenate(embeddings) 

  if its_colub:
      np.save('toxic_embeddings_8000', features)
  else:
      np.save('toxic_embeddings_2000', features)

else:
  if its_colub:
      features = np.load('./datasets/toxic_embeddings8000.npy')
  else:
      features = np.load('./datasets/toxic_embeddings2000.npy')

CPU times: user 1.1 ms, sys: 731 ms, total: 732 ms
Wall time: 2.02 s


<div class="alert alert-block alert-success">
<b>Успех:</b> Радует, что тебе удалось освоить БЕРТ!
</div>

In [11]:
targets = df['toxic']
features_train, features_valid, target_train, target_valid = train_test_split(features, targets, shuffle=True, test_size=0.25, random_state=12345)

<div class="alert alert-block alert-info">
<b>Совет: </b> В данном случае выборки лучше назвать train и test.
</div>

### Классификатор LogisticRegression

In [12]:
cv_counts = 3 #StratifiedKFold(n_splits=3, shuffle=True, random_state=12345)

In [13]:
%%time

classificator = LogisticRegression(max_iter=150, random_state=12345)
train_f1 = cross_val_score(classificator, 
                      features_train, 
                      target_train, 
                      cv=cv_counts, 
                      scoring='f1').mean()
print('F1 LogisticRegression', train_f1)

F1 LogisticRegression 0.46363115693012596
CPU times: user 3min 43s, sys: 13.8 s, total: 3min 56s
Wall time: 1min 26s


На несбалансированных классах логистическая регрессия дает результат, неудовлетворяющий условиям задачи. Попробуем балансировку через коэффициент и подбор гипермаркетов

In [15]:
dict_classes={0:1, 1:class_ratio}

In [ ]:
%%time

classificator = LogisticRegression(max_iter=100, class_weight=dict_classes, random_state=12345)
hyperparams = [{'solver':['newton-cg', 'lbfgs', 'liblinear'],
                'C':[0.1, 1, 10, 100]
              }]


print('# Tuning hyper-parameters for f1_score')
print()
clf = GridSearchCV(classificator, hyperparams, scoring='f1', cv=cv_counts, n_jobs=-1)
clf.fit(features_train, target_train)
print("Best parameters set found on development set:")
print()
LR_best_params = clf.best_params_
print(LR_best_params)
print()
print("Grid scores on development set:")
print()
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.6f for %r"% (mean, params))
print()

cv_f1_LR = max(means)
print('F1 на cv', cv_f1_LR)

In [16]:
%%time

LR_best_params = {'C': 10, 'class_weight': [dict_classes], 'solver': 'lbfgs'}

classificator = LogisticRegression(random_state=12345)
classificator.set_params(**LR_best_params)
classificator.fit(features_train, target_train)
target_predict = classificator.predict(features_valid)
valid_f1_LR = f1_score(target_valid, target_predict)

print('F1 на валидации', valid_f1_LR)

F1 на валидации 0.5263157894736843
CPU times: user 1min 16s, sys: 3.81 s, total: 1min 20s
Wall time: 26.4 s


Попробуем автоматическую балансировку

In [33]:
%%time

LR_best_params = {'C': 10, 'class_weight': 'balanced', 'solver': 'lbfgs'}

classificator = LogisticRegression(random_state=12345)
classificator.set_params(**LR_best_params)
classificator.fit(features_train, target_train)
target_predict = classificator.predict(features_valid)
valid_f1_LR = f1_score(target_valid, target_predict)

print('F1 на валидации', valid_f1_LR)

F1 на валидации 0.8131313131313133
CPU times: user 7min 43s, sys: 13min 12s, total: 20min 56s
Wall time: 44min 14s


In [34]:
print(confusion_matrix(target_valid, target_predict)) 

[[825 154]
 [216 805]]


Автобалансировка отрабатывает намного лучше. Результат уже достаточен длы выполнения задания. Сделаем resampling вручную и сравним

In [ ]:
df_train = df.iloc[target_train.index]

target_train_0 = df_train[df_train['toxic'] == 0]['toxic']
target_train_1 = df_train[df_train['toxic'] == 1]['toxic']

target_train_0_downsample = target_train_0.sample(target_train_1.shape[0], random_state=12345)
target_train_resample = pd.concat([target_train_0_downsample, target_train_1])

features_train_resample = features[target_train.index]
features_train, target_train = shuffle(features_train, target_train, random_state=12345)

<div class="alert alert-block alert-success">
<b>Успех:</b> Вот тут все верно, молодец!
</div>

Рассчитаем f1 предсказаний снова

In [ ]:
%%time

LR_best_params = {'C': 10, 'class_weight': 'balanced', 'solver': 'lbfgs'}

classificator = LogisticRegression(random_state=12345)
classificator.set_params(**LR_best_params)
classificator.fit(target_train_resample, features_train_resample)
target_predict = classificator.predict(features_valid)
valid_f1_LR = f1_score(target_valid, target_predict)

print('F1 на валидации', valid_f1_LR)

Результат хуже, чем автобалансировка - значит дальше будем работать с автобалансировкой


In [ ]:
del target_train_resample
del features_train_resample

### Классификатор DecisionTreeClassifier

In [ ]:
%%time

classificator = DecisionTreeClassifier(random_state=12345, class_weight='balanced')
hyperparams = [{'max_depth':[x for x in range(30,50,5)]
                }]


print('# Tuning hyper-parameters for f1_score')
print()
clf = GridSearchCV(classificator, hyperparams, scoring='f1', cv=cv_counts)
clf.fit(features_train, target_train)
print("Best parameters set found on development set:")
print()
DTC_best_params = clf.best_params_
print(DTC_best_params)
print()
print("Grid scores on development set:")
print()
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.6f for %r"% (mean, params))
print()

cv_f1_DTC = max(means)
print('F1 на cv', cv_f1_DTC)

In [35]:
%%time

DTC_best_params = {'class_weight': 'balanced', 'max_depth': 30, 'random_state': 12345}

classificator = DecisionTreeClassifier()
classificator.set_params(**DTC_best_params)
classificator.fit(features_train, target_train)
target_predict = classificator.predict(features_valid)
valid_f1_DTC = f1_score(target_valid, target_predict)

print('F1 на валидации', valid_f1_DTC)

F1 на валидации 0.6968058968058968
CPU times: user 18min 55s, sys: 21.2 s, total: 19min 16s
Wall time: 23min 3s


### Классификатор SGDClassifier

In [ ]:
%%time

classificator = SGDClassifier()
hyperparams = [{'loss':['hinge', 'log', 'modified_huber'],
                'learning_rate':['constant', 'optimal', 'invscaling', 'adaptive'],
                'eta0':[0.01, 0.1, 0.2, 0.5],
                'random_state':[12345],
                'class_weight': [dict_classes]
               }]


print('# Tuning hyper-parameters for f1_score')
print()
clf = GridSearchCV(classificator, hyperparams, scoring='f1', cv=cv_counts)
clf.fit(features_train, target_train)
print("Best parameters set found on development set:")
print()
SGDC_best_params = clf.best_params_
print(SGDC_best_params)
print()
print("Grid scores on development set:")
print()
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.6f for %r"% (mean, params))
print()

cv_f1_SGDC = max(means)
print('F1 на cv', cv_f1_SGDC)

In [ ]:
%%time

SGDC_best_params = {'class_weight': 'balanced', 'eta0': 0.2, 'learning_rate': 'adaptive', 'loss': 'modified_huber', 'random_state': 12345}

classificator = SGDClassifier()
classificator.set_params(**SGDC_best_params)
classificator.fit(features_train, target_train)
target_predict = classificator.predict(features_valid)
valid_f1_SGDC = f1_score(target_valid, target_predict)

print('F1 на валидации', valid_f1_SGDC)

F1 на валидации 0.8181818181818181
CPU times: user 3min 49s, sys: 7min 52s, total: 11min 42s
Wall time: 21min 42s


## Выводы

In [ ]:
index = ['LogisticRegression',
         'DecisionTreeClassifier',
         'SGDClassifier']
# data = {'f1 на Train':[cv_f1_LR,
#                     cv_f1_DTC,
#                     cv_f1_SGDC],
#         'f1 на Test':[valid_f1_LR,
#                            valid_f1_DTC,
#                            valid_f1_SGDC]}
data = {'f1 на Test':[valid_f1_LR,
                           valid_f1_DTC,
                           valid_f1_SGDC]}
scores_data = pd.DataFrame(data=data, index=index)
scores_data['Выполнение задачи'] = scores_data['f1 на Test'] > 0.75
display(scores_data)

f1 на Test:   
LogisticRegression     0.8131313131313133   
DecisionTreeClassifier 0.6968058968058968   
SGDClassifier          0.818181818181818   

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что ты попробовал несколько разных методов борьбы с дисбалансом классов и несколько моделей!
</div>

### Итого

В ходе реализации проекта сделано:

Подготовленны данные обучения типа Embeddings при помощи модели BERT.
Выбран способ баланса классов и поделены данные на обучающую, валидационную и тестовою выборку.
Обучены модели и выбраны лучшие из них на тестовой выборке.

В исходных даныых получается очень много фичей. Расчет Embeddings 2тыс записей на CPU длится 7 часов!!!! GridSearchCV считается 5-6 часов. На платном Colab 8 тыс записей BERT считается 1час

На тестовой выбоке по F1 лучше всего модель SGDClassifier. При этом результат незначительно превосходит LogisticRegression.

<div class="alert alert-block alert-success">
<b>Успех:</b> Приятно видеть подробный вывод в конце проекта!
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны